app/Http/Controllers/Table/LocationController.php

```diff
@@ -76,21 +85,32 @@ public function baseQuery($request)
      */
     public function formatItem($location)
     {
-        return [
+        $data = [
             'id' => $location->id,
             'location' => $location->location,
             'lat' => $location->lat,
             'lng' => $location->lng,
             'down' => $location->devices()->isDown()->count(),
             'devices' => $location->devices()->count(),
-            'network' => $location->devices()->where('type', 'network')->count(),
-            'servers' => $location->devices()->where('type', 'server')->count(),
-            'firewalls' => $location->devices()->where('type', 'firewall')->count(),
         ];
+        foreach (\LibreNMS\Config::get('device_types') as $device_type) {
```

Comment:
I think we can simplify all these queries a lot and drop the joins.  Eloquent now has a withCount method:
https://laravel.com/docs/8.x/eloquent-relationships#counting-related-models

Reply:
Nevermind, can't use that with groubBy

Reply:
Instead of many queries, we should do this with just one query:

```php
$counts = $location->devices()->groupBy('type')->select('type', DB::raw('COUNT(*) as 'total')->pluck('total', 'type');
foreach(\LibreNMS\Config::get('device_types') as $device_type) {
    $data[$device_type['type']] = $counts->get($device_type['type'], 0);
}
```